### PR TITLE
gh-118647: Add defaults to typing.Generator and typing.AsyncGenerator

### PR DESCRIPTION
**Pull request by JelleZijlstra**
* Issue: gh-118647

----
📚 Documentation preview 📚: https://cpython-previews--118648.org.readthedocs.build/

### CODE REVIEW

**Comment by Daverball** [comment]
It might be worth considering also updating `ContextManager` and `AsyncContextManager` to match what I did in https://github.com/python/typeshed/pull/11048 for `AbstractContextManager` and `AbstractAsyncContextManager`, now that you've laid the necessary groundwork for defaults in `_SpecialGenericAlias`.

But we can also save it for a follow-up PR, if you'd prefer that. A change in typeshed doesn't make sense until we test 3.13 to the CI.

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -3754,8 +3763,14 @@ Aliases to other ABCs in :mod:`collections.abc`
    of :class:`Generator` behaves contravariantly, not covariantly or
    invariantly.
 
-   If your generator will only yield values, set the ``SendType`` and
-   ``ReturnType`` to ``None``::
+   The ``SendType`` and ``ReturnType`` default to :const:`!None`::
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
   The ``SendType`` and ``ReturnType`` parameters default to :const:`!None`::
```

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -3754,8 +3763,14 @@ Aliases to other ABCs in :mod:`collections.abc`
...
+
+      def infinite_stream(start: int) -> Generator[int]:
+         while True:
+            yield start
+            start += 1
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
      def infinite_stream(start: int) -> Generator[int]:
          while True:
              yield start
              start += 1
```

**Review by AlexWaygood — APPROVED** [review]

**Comment by graingert** [comment]
`typing.Generator` and `typing.AsyncGenerator` are deprecated, why are they getting new features?

**Comment by mikeshardmind** [comment]
> `typing.Generator` and `typing.AsyncGenerator` are deprecated, why are they getting new features?

I would hope this isn't a real blocker here. From a user standpoint, users are told `collections.abc` variants are meant to behave identically. (they are actually documented as a deprecated *alias* ant not as a separate object)

> Deprecated alias to [collections.abc.Generator](https://docs.python.org/3/library/collections.abc.html#collections.abc.Generator).

If this deprecation results in them not being equivalent anymore, you've broken the documented idea of them being equivalent.

File: Lib/typing.py [hunk 3]
```diff
@@ -2813,8 +2824,8 @@ class Other(Leaf):  # Error reported by type checker
 OrderedDict = _alias(collections.OrderedDict, 2)
 Counter = _alias(collections.Counter, 1)
 ChainMap = _alias(collections.ChainMap, 2)
-Generator = _alias(collections.abc.Generator, 3)
-AsyncGenerator = _alias(collections.abc.AsyncGenerator, 2)
+Generator = _alias(collections.abc.Generator, 3, defaults=(type(None), type(None)))
```
Comment (sobolevn) [inline Lib/typing.py]:
```suggestion
Generator = _alias(collections.abc.Generator, 3, defaults=(types.NoneType, types.NoneType))
```

File: Lib/typing.py [hunk 4]
```diff
@@ -2813,8 +2824,8 @@ class Other(Leaf):  # Error reported by type checker
...
+AsyncGenerator = _alias(collections.abc.AsyncGenerator, 2, defaults=(type(None),))
```
Comment (sobolevn) [inline Lib/typing.py]:
```suggestion
AsyncGenerator = _alias(collections.abc.AsyncGenerator, 2, defaults=(types.NoneType,))
```

**Review by sobolevn — APPROVED** [review]

**Pull request merged**